src/app/clusters/concentration-measurement-server/concentration-measurement-server.cpp

```diff
@@ -0,0 +1,512 @@
+    mPeakMeasuredValueWindow = aPeakMeasuredValueWindow;
+
+    if (oldValue != mPeakMeasuredValueWindow)
+    {
+        ConcreteAttributePath path = ConcreteAttributePath(mEndpointId, mClusterId, Attributes::PeakMeasuredValueWindow::Id);
+        MatterReportingAttributeChangeCallback(path);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Instance::SetAverageMeasuredValue(DataModel::Nullable<float> aAverageMeasuredValue)
+{
+    if (!HasFeature(Feature::kAverageMeasurement))
+    {
+        return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+    }
+
+    if (!aAverageMeasuredValue.IsNull())
+    {
+        if (!mMaxMeasuredValue.IsNull())
+        {
+            if (aAverageMeasuredValue.Value() > mMaxMeasuredValue.Value())
+            {
+                return CHIP_ERROR_INVALID_ARGUMENT;
+            }
+        }
+
+        if (!mMinMeasuredValue.IsNull())
+        {
+            if (aAverageMeasuredValue.Value() < mMinMeasuredValue.Value())
+            {
+                return CHIP_ERROR_INVALID_ARGUMENT;
+            }
+        }
+    }
+
+    DataModel::Nullable<float> oldValue = mAverageMeasuredValue;
+    mAverageMeasuredValue               = aAverageMeasuredValue;
+
+    if (oldValue != mAverageMeasuredValue)
+    {
+        ConcreteAttributePath path = ConcreteAttributePath(mEndpointId, mClusterId, Attributes::AverageMeasuredValue::Id);
+        MatterReportingAttributeChangeCallback(path);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Instance::SetAverageMeasuredValueWindow(uint32_t aAverageMeasuredValueWindow)
+{
+    if (!HasFeature(Feature::kAverageMeasurement))
+    {
+        return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+    }
+
+    if (aAverageMeasuredValueWindow > WINDOW_MAX)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    uint32_t oldValue           = mAverageMeasuredValueWindow;
+    mAverageMeasuredValueWindow = aAverageMeasuredValueWindow;
+
+    if (oldValue != mAverageMeasuredValueWindow)
+    {
+        ConcreteAttributePath path = ConcreteAttributePath(mEndpointId, mClusterId, Attributes::AverageMeasuredValueWindow::Id);
+        MatterReportingAttributeChangeCallback(path);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Instance::SetUncertainty(float aUncertainty)
+{
+    if (!HasFeature(Feature::kNumericMeasurement))
+    {
+        return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+    }
+
+    float oldValue = mUncertainty;
+    mUncertainty   = aUncertainty;
+
+    if (oldValue != mUncertainty)
+    {
+        ConcreteAttributePath path = ConcreteAttributePath(mEndpointId, mClusterId, Attributes::Uncertainty::Id);
+        MatterReportingAttributeChangeCallback(path);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Instance::SetLevel(LevelValueEnum aLevel)
+{
+    if (!HasFeature(Feature::kLevelIndication))
+    {
+        return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+    }
+
+    if (aLevel == LevelValueEnum::kUnknownEnumValue)
```

Comment:
This test does not seem to be what you want.  This value is not coming off the wire, where we guarantee that we convert all unknown values to kUnknownEnumValue.  It could have all sorts of garbage values.  If we actually are trying to do this check to catch people casting random invalid things to LevelValueEnum, this probably needs to be something similar to what EnsureKnownEnumValue bits do....